hera_cal/abscal_funcs.py

```diff
@@ -0,0 +1,1496 @@
+"""
+abscal_funcs.py
+---------------
+
+Self contained routines
+used by abscal.py
+"""
+import os
+import sys
+from collections import OrderedDict as odict
+import copy
+import argparse
+import functools
+import numpy as np
+from pyuvdata import UVCal, UVData
+from pyuvdata import utils as uvutils
+from hera_cal import omni, utils, firstcal, cal_formats
+from scipy import signal
+from scipy import interpolate
+import linsolve
+import itertools
+import operator
+
+
+def abs_amp_logcal(model, data, wgts=None, verbose=True):
+    """
+    calculate absolute (array-wide) gain amplitude scalar
+    with a linear solver using logarithmically linearized equation:
+
+    ln|V_ij,xy^model / V_ij,xy^data| = eta_x + eta_y
+
+    where {i,j} index antenna numbers and {x,y} index polarizations
+
+    Parameters:
+    -----------
+    model : visibility data of refence model, type=dictionary
+            keys are antenna-pair + polarization tuples, Ex. (1, 2, 'xx').
+            values are complex ndarray visibilities.
+            these must be 2D arrays, with [0] axis indexing time
+            and [1] axis indexing frequency.
+
+    data : visibility data of measurements, type=dictionary
+           keys are antenna pair + poltuples (must match model), values are
+           complex ndarray visibilities matching shape of model
+
+    wgts : weights of data, type=dictionry, [default=None]
+           keys are antenna pair + poltuples (must match model), values are real floats
+           matching shape of model and data
+
+    verbose : print output, type=boolean, [default=False]
+
+    Output:
+    -------
+    fit : dictionary with 'amp_x' key for amplitude scalar for x polarization,
+            which has the same shape as the ndarrays in the model
+    """
+    echo("...configuring linsolve data for abs_amp_logcal", verbose=verbose)
+
+    # get keys from model dictionary
+    keys = sorted(set(model.keys()) & set(data.keys()))
+
+    # abs of amplitude ratio is ydata independent variable
+    ydata = odict([(k, np.log(np.abs(model[k]/data[k]))) for k in keys])
+
+    # make weights if None
+    if wgts is None:
+        wgts = copy.deepcopy(ydata)
```

Comment:
Maybe do this as a dictionary comprehension instead to not copy all data. Possibility for bloating memory?